File: golang-alg/search/binarySearch.go
package main

import (
	"fmt"
)

func main() {
	arr := []int{1, 2, 3, 4, 5, 6, 7, 8, 9, 10}
	fmt.Println(binarySearch(arr, 3))
}

// ===
// 前提是排好序的array
// ===
func binarySearch(arr []int, n int) int {
	if len(arr) <= 1 {
		return 0
	}
	left := 0
	right := len(arr) - 1
	for left <= right {
		middle := (left + right) / 2
		if arr[middle] > n {
			right = middle - 1
			fmt.Println("right", right)
		} else if arr[middle] < n {
			left = middle + 1
			fmt.Println("left", left)
		} else {
			fmt.Println("middle", middle)
			return middle
		}
	}
	return -1
}

File: python-alg/basic/binary_search.py
def binary_search(arr, target):
    left = 0
    right = len(arr) - 1

    while left <= right:
        mid = left + (right - left) // 2
        if arr[mid] == target:
            return mid
        elif arr[mid] < target:
            left = mid + 1
        else:
            right = mid - 1

    return -1


my_list = [3, 9, 10, 27, 38, 43, 82]
index = binary_search(my_list, 43)
print(index)
